### PR TITLE
Improving Performance and creation of intances

### PR DESCRIPTION
**Pull request by ChicK00o**
Have a look and let me know

### CODE REVIEW

File: DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+/// This extension is a benchmarked better implementation to Activator.CreateInstance when taking the second link into consideration
+/// Here the expression is further cached so as to better instantiation for factory implementations
+/// </summary>
+internal static class InstanceCreatorExtensions
+{
+    internal static object New(this Type type, params object[] args)
+    {
+        Type[] argTypes = args.Select(a => a.GetType()).ToArray();
+
+        var ctor = type.GetConstructor(argTypes);
```
Comment (disturbing) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs]: what is ctro - try to use full names and define variables versus using var.

**Comment by ChicK00o** [comment]
Have a look again

File: DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs (hunk 1 above)
Comment (ChicK00o) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs]: My bad :)

File: DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        return GetActivator(ctor)(args);
+    }
+
+    internal delegate object ObjectActivator(params object[] args);
+
+    private static Dictionary<ConstructorInfo, object> _activators = new Dictionary<ConstructorInfo, object>();
```
Comment (disturbing) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs]: Move dictionary to top of class and delegate definition as per code format rules for this project.

Comment (disturbing) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs]: remove _ as private variable - this project doesn't do that in terms of format.

File: DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    internal static ObjectActivator GetActivator(ConstructorInfo ctor)
+    {
+        object activator;
+        if (_activators.TryGetValue(ctor, out activator)) return (ObjectActivator)activator;
```
Comment (disturbing) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs]: No need to store in dictionary, object - you can store in dictionary, ObejctActivator, so we don't need to have a cast here.
- Also for this project I've tried to have single return out per function rule Please reformat to do the following:

```
ObjectActivator activator;

if (TryGetValue == false){
     activator = CreateNewActivator(params);
}

return activator;


```

File: DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        //create a single param of type object[]
+        ParameterExpression param = Expression.Parameter(typeof(object[]), "args");
+
+        var argsExp = new Expression[paramsInfo.Length];
+
+        //pick each arg from the params array
+        //and create a typed expression of them
+        for (int i = 0; i < paramsInfo.Length; i++)
+        {
```
Comment (disturbing) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs:52]: Same line brackets.

File: DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    internal static ObjectActivator GetActivator(ConstructorInfo ctor)
+    {
+        object activator;
+        if (_activators.TryGetValue(ctor, out activator)) return (ObjectActivator)activator;
+
+        ParameterInfo[] paramsInfo = ctor.GetParameters();
+
+        //create a single param of type object[]
+        ParameterExpression param = Expression.Parameter(typeof(object[]), "args");
```
Comment (disturbing) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs]: use constant for name 'ARGS'.

File: DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        }
+
+        //make a NewExpression that calls the ctor with the args we just created
+        NewExpression newExp = Expression.New(ctor, argsExp);
+
+        //create a lambda with the NewExpression as body and our param object[] as arg
+        LambdaExpression lambda = Expression.Lambda(typeof(ObjectActivator), newExp, param);
+
+        //compile it
+        var compiled = (ObjectActivator)lambda.Compile();
```
Comment (disturbing) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs]: static defined types, remove var please.

File: DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+/// <summary>
+/// Got this from the following sites :
+/// https://vagifabilov.wordpress.com/2010/04/02/dont-use-activator-createinstance-or-constructorinfo-invoke-use-compiled-lambda-expressions/
+/// https://ayende.com/blog/3167/creating-objects-perf-implications
+/// http://rogeralsing.com/2008/02/28/linq-expressions-creating-objects/
+/// https://rboeije.wordpress.com/2012/04/19/extension-method-on-type-as-alternative-for-activator-createinstance/ - Implemented this
+/// 
+/// This extension is a benchmarked better implementation to Activator.CreateInstance when taking the second link into consideration
+/// Here the expression is further cached so as to better instantiation for factory implementations
```
Comment (disturbing) [inline DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs]: Add your github name as @author

**Comment by disturbing** [comment]
Sorry, I was not able to finish the full review until now.

I just wrote the remaining comments.  Once all comments are resolved, I will accept the pull request.

Great work overall.

**Comment by ChicK00o** [comment]
Incorporated the changes you mentioned, have a look again and
Let me know if there are any further changes you need
Cheers

**Comment by disturbing** [comment]
All good :+1:

**Pull request merged**